airbyte-integrations/connectors/source-zuora/source_zuora/zuora_client.py

```diff
@@ -0,0 +1,275 @@
+        # If there is no `end_date` as input - use now()
+        e = pendulum.parse(pendulum.now().to_datetime_string()) if end_date is None else pendulum.parse(end_date)
+        # check end_date bigger than start_date
+        if s >= e:
+            raise EndDateError
+        else:
+            # Get n of date-slices
+            n = self._get_n_slices(s, e, window_days)
+            # initiating slice_start/end for date slice
+            slice_start = s.to_datetime_string() + tz
+            # if slice_end is bigger than the input end_date, switch to the end_date
+            slice_end = self._check_end_date(self._next_slice_end_date(slice_start, window_days), e) + tz
+            # For multiple date-slices
+            if n > 1:
+                while n > 0:
+                    yield from self._get_data(q_type, obj, date_field, slice_start, slice_end)
+                    # make next date-slice
+                    slice_start = slice_end
+                    # if next date-slice end_date is bigger than the input end_date, we switch to the end_date
+                    slice_end = self._check_end_date(self._next_slice_end_date(slice_end, window_days), e) + tz
+                    # Modify iterator
+                    n -= 1
+            else:
+                # For 1 date-slice
+                yield from self._get_data(q_type, obj, date_field, slice_start, slice_end)
+
+    # Check the end-date before assign next date-slice
+    @staticmethod
+    def _check_end_date(slice_end_date, global_end_date: str) -> str:
+        return global_end_date.to_datetime_string() if pendulum.parse(slice_end_date) > global_end_date else slice_end_date
+
+    # Get n of date-slices needed to fetch the data
+    @staticmethod
+    def _get_n_slices(start_date: datetime, end_date: datetime, window_days: int) -> int:
+        # case where we have 1 sec difference between start and end dates would produce 0 slices, so return 1 instead
+        d = pendulum.period(start_date, end_date).in_days()
+        return 1 if ceil(d / window_days) <= 0 else ceil(d / window_days)
+
+    # Creates the next-date-slice
+    @staticmethod
+    def _next_slice_end_date(end_date: str, window_days: int) -> str:
+        # potentially could be replaced with next_page_token
+        format = "YYYY-MM-DD HH:mm:ss Z"
+        return pendulum.from_format(end_date, format).add(days=window_days).to_datetime_string()
+
+    # Get/Set the timezone information from the date_field
+    @staticmethod
+    def _get_tz(start_date: datetime) -> str:
+        return " +00:00" if start_date.tzname() == "UTC" else f" {start_date.tzname()}"
+
+    # Compares strings for datatypes casting
+    @staticmethod
+    def _check_data_type(_str, _list):
+        # Returns True if the Type is in the List of Types, else - False
+        return any(_str in s for s in _list)
+
+    # Convert Zuora DataTypes to JsonSchema Types
+    def _cast_schema_types(self, _schema: List) -> Dict:
```

Comment:
could you create unit test for this method ?

Reply:
On my way of doing so.

Reply:
still no tests

Reply:
Added:
-  unit_test for `convert_schema_types`
- 2 acceptance tests: `test_zuora_list_streams`, `test_zuora_get_json_schema`